core/src/main/java/org/apache/iceberg/MetricsConfig.java

```diff
@@ -107,6 +115,81 @@ public static MetricsConfig forPositionDelete(Table table) {
     return new MetricsConfig(columnModes.build(), defaultMode);
   }
 
+  static Iterable<Integer> breadthFirstFieldPriority(Schema schema) {
+    return TypeUtil.visit(
+        schema,
+        new TypeUtil.CustomOrderSchemaVisitor<>() {
+
+          @Override
+          public Iterable<Integer> schema(Schema schema, Supplier<Iterable<Integer>> structResult) {
+            return structResult.get();
+          }
+
+          @Override
+          public Iterable<Integer> struct(
+              Types.StructType struct, Iterable<Iterable<Integer>> fieldResults) {
+
+            List<Integer> orderedFieldIds =
+                Lists.newArrayListWithExpectedSize(struct.fields().size());
+            for (Types.NestedField field : struct.fields()) {
+              if (field.type().isPrimitiveType()) {
+                orderedFieldIds.add(field.fieldId());
+              }
+            }
+            return Iterables.concat(orderedFieldIds, Iterables.concat(fieldResults));
+          }
+
+          @Override
+          public Iterable<Integer> field(
+              Types.NestedField field, Supplier<Iterable<Integer>> future) {
+            return future.get();
+          }
+
+          @Override
+          public Iterable<Integer> list(Types.ListType list, Supplier<Iterable<Integer>> future) {
+            List<Integer> returnValue = Lists.newArrayListWithCapacity(1);
+            if (list.elementType().isPrimitiveType()) {
```

Comment:
I think that this can be simpler:

```java
  if (list.elementType.isPrimitiveType() || list.elementType().isVariantType()) {
    return ImmutableList.of(list.elementId());
  }

  return future.get();
```

If the element is a primitive type (or a variant), then we don't need to get the value of the future because we already know that it is going to be `Collections.emptyList()`. I prefer that implementation because it cuts down on `Iterable` instances that concatenate things that we know are empty.